File: day1/d1.c
#include "ctype.h"
#include "stdio.h"
#include "string.h"

const char* numbers[] =
{
  "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
};

int get_digit_p1(char* p)
{
  if (isdigit(p[0])) return p[0] - '0';
  return 0;
}

int get_digit_p2(char* p)
{
  int p1 = get_digit_p1(p);
  if (p1) return p1;

  for (int i = 0; i < 9; i++)
  {
    if (strncmp(p, numbers[i], strlen(numbers[i])) == 0) return i + 1;
  }

  return 0;
}

int main()
{
  FILE* f = fopen("input.txt", "r");
  char buf[100];
  int total[2] = {0, 0};
  while (fscanf(f, "%s\n", buf) != EOF)
  {
    int digits_p1[2] = {0, 0};
    int digits_p2[2] = {0, 0};
    size_t len = strlen(buf);
    for (int i = 0; i < len; i++)
    {
      if (!digits_p1[0]) digits_p1[0] = get_digit_p1(&buf[i]);
      if (!digits_p2[0]) digits_p2[0] = get_digit_p2(&buf[i]);
      if (!digits_p1[1]) digits_p1[1] = get_digit_p1(&buf[len - i - 1]);
      if (!digits_p2[1]) digits_p2[1] = get_digit_p2(&buf[len - i - 1]);
    }

    total[0] += 10 * digits_p1[0] + digits_p1[1];
    total[1] += 10 * digits_p2[0] + digits_p2[1];
  }

  printf("P1: %d, P2: %d\n", total[0], total[1]);

  return 0;
}

File: day01/d1.zig
const std = @import("std");

const numbers: [9][]const u8 = [_][]const u8{
  "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
};

pub fn get_digit_p1(buf: []u8) u32 {
  if (std.ascii.isDigit(buf[0])) return buf[0] - '0';
  return 0;
}

pub fn get_digit_p2(buf: []u8) u32 {
  var p1: u32 = get_digit_p1(buf);
  if (p1 != 0) return p1;

  for (numbers, 0..) |n, i| {
    if (std.mem.startsWith(u8, buf, n)) return @as(u32, @intCast(i + 1));
  }

  return 0;
}

pub fn main() !void
{
  const f = try std.fs.cwd().openFile("input.txt", .{ .mode = std.fs.File.OpenMode.read_only });
  defer f.close();

  var buf: [100]u8 = undefined;
  var totals = [2]u32{0, 0};
  while (try f.reader().readUntilDelimiterOrEof(&buf, '\n')) |line| {
    var digits1 = [2]u32{0, 0};
    var digits2 = [2]u32{0, 0};
    for (line, 0..) |_, i| {
      if (digits1[0] == 0) digits1[0] = get_digit_p1(line[i..line.len]);
      if (digits1[1] == 0) digits1[1] = get_digit_p1(line[line.len-i-1..line.len]);
      if (digits2[0] == 0) digits2[0] = get_digit_p2(line[i..line.len]);
      if (digits2[1] == 0) digits2[1] = get_digit_p2(line[line.len-i-1..line.len]);
    }
    totals[0] += 10 * digits1[0] + digits1[1];
    totals[1] += 10 * digits2[0] + digits2[1];
  }
  const stdout = std.io.getStdOut().writer();
  try stdout.print("P1: {d}, P2: {d}\n", .{totals[0], totals[1]});
}
